observer.cpp

```diff
@@ -87,5 +87,38 @@ void test_observer()
         assert(o.get_params() == params);
     }
 #endif
+
+//#define FIX_ISSUE_81
+#ifdef FIX_ISSUE_81
+  ///The observer stores inputs and optimal values
+  {
+    observer o;
+
+    //Give sim some non-default inputs and optimal
+    simulation s{};
+    int n_repeats = 100;
+
+    for(int i = 0; i != n_repeats; ++i){
+        tick(s);
+        assert(o.get_input()[i] != s.get_input());
+        assert(o.get_optimal()[i] != s.get_optimal());
+
+        o.store_input(s);
+        o.store_optimal(s);
+
+        assert(o.get_input()[i] == s.get_input());
+        assert(o.get_optimal()[i] == s.get_optimal());
```

Comment:
here you can use [i], but it is clearer and neater if you use `.back()`, as in any casy you always check the last element of the `vector`s.
```suggestion
        assert(o.get_input().back() == s.get_input());
        assert(o.get_optimal().back() == s.get_optimal());
```